src/infra/dex/okx/mod.rs

```diff
@@ -0,0 +1,274 @@
+            },
+        })
+    }
+
+    pub async fn swap(
+        &self,
+        order: &dex::Order,
+        slippage: &dex::Slippage,
+    ) -> Result<dex::Swap, Error> {
+        let query = self
+            .defaults
+            .clone()
+            .with_domain(order, slippage)
+            .ok_or(Error::OrderNotSupported)?;
+        let quote = {
+            // Set up a tracing span to make debugging of API requests easier.
+            // Historically, debugging API requests to external DEXs was a bit
+            // of a headache.
+            static ID: AtomicU64 = AtomicU64::new(0);
+            let id = ID.fetch_add(1, atomic::Ordering::Relaxed);
+            self.quote(&query)
+                .instrument(tracing::trace_span!("quote", id = %id))
+                .await?
+        };
+
+        Self::handle_api_error(quote.code, &quote.msg)?;
+        let quote_result = quote.data.first().ok_or(Error::NotFound)?;
+
+        // Increasing returned gas by 50% according to the documentation:
+        // https://www.okx.com/en-au/web3/build/docs/waas/dex-swap (gas field description in Response param)
+        let gas = quote_result
+            .tx
+            .gas
+            .checked_add(quote_result.tx.gas >> 1)
+            .ok_or(Error::GasCalculationFailed)?;
+
+        Ok(dex::Swap {
+            call: dex::Call {
+                to: eth::ContractAddress(quote_result.tx.to),
+                calldata: quote_result.tx.data.clone(),
+            },
+            input: eth::Asset {
+                token: quote_result
+                    .router_result
+                    .from_token
+                    .token_contract_address
+                    .into(),
+                amount: quote_result.router_result.from_token_amount,
+            },
+            output: eth::Asset {
+                token: quote_result
+                    .router_result
+                    .to_token
+                    .token_contract_address
+                    .into(),
+                amount: quote_result.router_result.to_token_amount,
+            },
+            allowance: dex::Allowance {
+                spender: eth::ContractAddress(quote_result.tx.to),
+                amount: dex::Amount::new(quote_result.router_result.from_token_amount),
+            },
+            gas: eth::Gas(gas),
+        })
+    }
+
+    async fn quote(&self, query: &dto::SwapRequest) -> Result<dto::SwapResponse, Error> {
+        let mut request_builder = self
+            .client
+            .request(reqwest::Method::GET, self.endpoint.clone())
+            .query(query);
+
+        let request = request_builder
+            .try_clone()
+            .ok_or(Error::RequestBuildFailed)?
+            .build()
+            .map_err(|_| Error::RequestBuildFailed)?;
+
+        let timestamp = &chrono::Utc::now()
+            .to_rfc3339_opts(SecondsFormat::Millis, true)
+            .to_string();
+        let signature = self.sign_request(&request, timestamp)?;
```

Comment:
nit: that function name would imply to me that you put in a request which gets signed and returned.
```suggestion
        let signature = self.generate_signature(&request, timestamp)?;
```